nsot/models.py

```diff
@@ -1470,6 +1494,131 @@ def to_dict(self):
+        null=False,
+        related_name='circuit_a',
+        verbose_name='A-side Interface',
+        help_text='Unique ID of interface at the A-side'
+    )
+
+    # z_endpoint interface
+    z_endpoint = models.OneToOneField(
+        Interface,
+        on_delete=models.CASCADE,
+        db_index=True,
+        null=True,
+        related_name='circuit_z',
+        verbose_name='Z-side Interface',
+        help_text='Unique ID of interface at the Z-side'
+    )
+
+    # We are currently inferring the site_id from the parent (A-side) Interface
+    # in the .save() method
+    site = models.ForeignKey(
+        Site, db_index=True, related_name='circuits',
+        on_delete=models.PROTECT,
+        help_text='Unique ID of the Site this Circuit is under'
+    )
+
+    name = models.CharField(
+        max_length=255, unique=True,
+        help_text='Unique display name of circuit',
+        default='',
+    )
+
+    def __unicode__(self):
+        return u'%s' % self.name
+
+    @property
+    def interfaces(self):
+        intf_list = [self.a_endpoint, self.z_endpoint]
+        intf_list = [i for i in intf_list if i]
+        return intf_list
+
+    @property
+    def addresses(self):
+        interfaces = self.interfaces
+        addresses = [
+            addr
+            for intf in interfaces
+            for addr in intf.addresses.all()
+        ]
+        return addresses
+
+    @property
+    def devices(self):
+        devices = [intf.device for intf in self.interfaces]
+        return devices
+
+    def clean_site(self, value):
+        """Always enforce that site is set."""
+        if value is None:
+            return self.a_endpoint.site_id
+
+        return value
+
+    def clean_a_endpoint(self, value):
+        if Circuit.objects.filter(z_endpoint=value).exists():
+            raise exc.ValidationError({
+                'a_endpoint': 'Interface already used as a z_endpoint'
+            })
+
+        return self.a_endpoint
+
+    def clean_z_endpoint(self, value):
+        if Circuit.objects.filter(a_endpoint=value).exists():
+            raise exc.ValidationError({
+                'z_endpoint': 'Interface already used as a a_endpoint'
+            })
+
+        return self.z_endpoint
+
+    def set_name(self):
+        if self.name:
+            return False
+
+        # Add display name of hostname:intf::hostname:intf
+        a_name = '{}:{}'.format(
+            self.a_endpoint.device.hostname,
+            self.a_endpoint.name,
+        )
+        if getattr(self, 'z_endpoint'):
+            z_name = '{}:{}'.format(
+                self.z_endpoint.device.hostname,
+                self.z_endpoint.name,
+            )
+        else:
+            z_name = 'None'
+
+        self.name = '{}_{}'.format(a_name, z_name)
+        return True
+
+    def clean_fields(self, exclude=None):
+        self.site_id = self.clean_site(self.site_id)
+        self.a_endpoint = self.clean_a_endpoint(self.a_endpoint_id)
```

Comment:
`self.name` should be set here so that when a `.full_clean() is performed, the name gets set.

Otherwise this happens (name isn't set):

```python
In [18]: c1 = Circuit.objects.create(a_endpoint=if1, z_endpoint=if2)

In [19]: c1
Out[19]: <Circuit: >
```

Which means the other places where `.set_name()` is called in a `CircuitSerializer.create()` and `CircuitSerializer.update()` can be removed and just rely on validation to cascade correctly.